src/detect-http2.c

```diff
@@ -0,0 +1,178 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Philippe Antoine <p.antoine@catenacyber.fr>
+ *
+ */
+
+#include "suricata-common.h"
+
+#include "detect.h"
+#include "detect-parse.h"
+#include "detect-engine.h"
+
+#include "detect-http2.h"
+#include "util-byte.h"
+#include "rust.h"
+
+/* prototypes */
+static int DetectHTTP2frametypeMatch(DetectEngineThreadCtx *det_ctx,
+                                     Flow *f, uint8_t flags, void *state, void *txv, const Signature *s,
+                                     const SigMatchCtx *ctx);
+static int DetectHTTP2frametypeSetup (DetectEngineCtx *, Signature *, const char *);
+void DetectHTTP2frametypeFree (void *);
+#ifdef UNITTESTS
+void DetectHTTP2RegisterTests (void);
+#endif
+
+static int g_http2_match_buffer_id = 0;
+
+static int DetectEngineInspectHTTP2(ThreadVars *tv, DetectEngineCtx *de_ctx,
+                                   DetectEngineThreadCtx *det_ctx, const Signature *s, const SigMatchData *smd,
+                                   Flow *f, uint8_t flags, void *alstate, void *txv, uint64_t tx_id)
+{
+    return DetectEngineInspectGenericList(tv, de_ctx, det_ctx, s, smd,
+                                          f, flags, alstate, txv, tx_id);
+}
+
+/**
+ * \brief Registration function for HTTP2 keywords
+ */
+
+void DetectHttp2Register(void)
+{
+    sigmatch_table[DETECT_HTTP2_FRAMETYPE].name = "http2.frametype";
+    sigmatch_table[DETECT_HTTP2_FRAMETYPE].desc = "match on HTTP2 frame type field";
+    //TODO create a new doc file for HTTP2 keywords
+    sigmatch_table[DETECT_HTTP2_FRAMETYPE].url = DOC_URL DOC_VERSION "/rules/http2-keywords.html#frametype";
+    sigmatch_table[DETECT_HTTP2_FRAMETYPE].Match = NULL;
+    sigmatch_table[DETECT_HTTP2_FRAMETYPE].AppLayerTxMatch = DetectHTTP2frametypeMatch;
+    sigmatch_table[DETECT_HTTP2_FRAMETYPE].Setup = DetectHTTP2frametypeSetup;
+    sigmatch_table[DETECT_HTTP2_FRAMETYPE].Free = DetectHTTP2frametypeFree;
+#ifdef UNITTESTS
+    sigmatch_table[DETECT_HTTP2_FRAMETYPE].RegisterTests = DetectHTTP2RegisterTests;
+#endif
+    //TODO investigate why this is needed
```

Comment:
@victorjulien can you point me to some explanation what is going on here ?

Reply:
Not sure I understand what you're asking.

Reply:
What does `DetectAppLayerInspectEngineRegister` do ?

Reply:
It registers an inspection engine to the statement part of the detection engine, for a specific protocol at a specific progress level.

Reply:
Thanks, what is "the statement part" ?

Reply:
Maybe it's best to look at an example from ssh (detect-ssh-software-version.c)

It first creates an inspect engine to be called when the SSH parser reaches the 'banner' state:

```c
    g_ssh_banner_list_id = DetectBufferTypeRegister("ssh_banner");

    DetectAppLayerInspectEngineRegister("ssh_banner",
            ALPROTO_SSH, SIG_FLAG_TOSERVER, SSH_STATE_BANNER_DONE,
            InspectSshBanner);
```
The inspection function is a generic one that just executes a normal 'SigMatch' list:
```c
static int InspectSshBanner(ThreadVars *tv,
        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
        const Signature *s, const SigMatchData *smd,
        Flow *f, uint8_t flags, void *alstate,
        void *txv, uint64_t tx_id)
{
    return DetectEngineInspectGenericList(tv, de_ctx, det_ctx, s, smd,
                                          f, flags, alstate, txv, tx_id);
}
```
(this wrapper may actually be unnecessary, could register `DetectEngineInspectGenericList` directly I think)

These things don't do anything by themselves yet, we now need one or more keywords that will need to evaluated for SSH, when the banner is available. This is what is done here for the 'ssh.software_version' keyword:

```c
static int DetectSshSoftwareVersionSetup (DetectEngineCtx *de_ctx, Signature *s, const char *str)
{
    DetectSshSoftwareVersionData *ssh = NULL;
    SigMatch *sm = NULL;

    if (DetectSignatureSetAppProto(s, ALPROTO_SSH) != 0)
        return -1;

    ssh = DetectSshSoftwareVersionParse(str);
    if (ssh == NULL)
        goto error;

    /* Okay so far so good, lets get this into a SigMatch
     * and put it in the Signature. */
    sm = SigMatchAlloc();
    if (sm == NULL)
        goto error;

    sm->type = DETECT_AL_SSH_SOFTWAREVERSION;
    sm->ctx = (void *)ssh;

    SigMatchAppendSMToList(s, sm, g_ssh_banner_list_id);
    return 0;
}
```

The last thing the setup function here does before returning, is adding the `SigMatch` to the banner inspection list.

Reply:
Many thanks for the walk through explanation :-)